format/applebookmark/apple_bookmark.go

```diff
@@ -0,0 +1,455 @@
+		case dataTypeNumber32:
+			d.FieldS32("data")
+		case dataTypeNumber64:
+			d.FieldS64("data")
+		case dataTypeNumber32F:
+			d.FieldF32("data")
+		case dataTypeNumber64F:
+			d.FieldF64("data")
+		case dataTypeDate:
+			d.FieldF64BE("data", scalar.DescriptionTimeFn(scalar.S.TryActualF, cocoaTimeEpochDate, time.RFC3339))
+		case dataTypeBooleanFalse:
+		case dataTypeBooleanTrue:
+		case dataTypeArray:
+			d.FieldStructNArray("data", "element", int64(n/arrayEntrySize), func(d *decode.D) {
+				offset := calcOffset(d.FieldU32("offset"))
+				d.SeekAbs(offset, decodeRecord)
+			})
+		case dataTypeDictionary:
+			d.FieldStructNArray("data", "element", int64(n/dictEntrySize), func(d *decode.D) {
+				keyOffset := calcOffset(d.FieldU32("key_offset"))
+				d.FieldStruct("key", func(d *decode.D) {
+					d.SeekAbs(keyOffset, decodeRecord)
+				})
+
+				valueOffset := calcOffset(d.FieldU32("value_offset"))
+				d.FieldStruct("value", func(d *decode.D) {
+					d.SeekAbs(valueOffset, decodeRecord)
+				})
+			})
+		case dataTypeUUID:
+			d.FieldRawLen("data", int64(n*8))
+		case dataTypeURL:
+			d.FieldUTF8("data", n)
+		case dataTypeRelativeURL:
+			baseOffset := d.FieldU32("base_url_offset")
+			d.FieldStruct("base_url", func(d *decode.D) {
+				d.SeekAbs(int64(baseOffset), decodeRecord)
+			})
+
+			suffixOffset := d.FieldU32("suffix_offset")
+			d.FieldStruct("suffix", func(d *decode.D) {
+				d.SeekAbs(int64(suffixOffset), decodeRecord)
+			})
+		}
+	})
+}
+
+const reservedSize = 32
+const headerEnd = 48
+const headerEndBitPos = headerEnd * 8
+
+// all offsets are calculated relative to the end of the bookmark header
+func calcOffset(i uint64) int64 { return int64(8 * (i + headerEnd)) }
+
+func bookmarkDecode(d *decode.D, _ any) any {
+	// all fields are little-endian with the exception of the Date datatype.
+	d.Endian = decode.LittleEndian
+
+	// decode bookmarkdata header, one at the top of each "file",
+	// although these may be nested inside of binary plists
+	d.FieldStruct("header", func(d *decode.D) {
+		d.FieldUTF8("magic", 4, d.AssertStr("book", "alis"))
+		d.FieldU32("total_size")
+		d.FieldU32("unknown")
+		d.FieldU32("header_size", d.AssertU(48))
+		d.FieldRawLen("reserved", reservedSize*8)
+	})
+
+	tocOffset := calcOffset(d.FieldU32("first_toc_offset"))
+
+	var currentHdr *tocHeader
+	var tocHeaders []*tocHeader
+
+	hdrCount := 0
+	d.FieldArrayLoop("toc_headers", func() bool {
+		return tocOffset != headerEndBitPos || hdrCount > 100
+	}, func(d *decode.D) {
+		d.SeekAbs(tocOffset, func(d *decode.D) {
+			currentHdr = decodeTOCHeader(d)
+			tocOffset = calcOffset(currentHdr.nextTOCOffset)
+			tocHeaders = append(tocHeaders, currentHdr)
+		})
+		hdrCount++
+	})
+
+	// now that we've collected all toc headers, iterate through each one's
+	// entries and decode associated records.
+	d.FieldArray("bookmark_entries",
+		func(d *decode.D) {
+			for _, hdr := range tocHeaders {
+				d.SeekAbs(hdr.entryArrayOffset, hdr.decodeEntries)
```

Comment:
Was confused how number of entries was pass to decodeEntries, didn't notice decodeEntries was a method, nice :)